core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml

```diff
@@ -4416,6 +4416,90 @@ LogicalProject(ENAME=[$1])
+  LogicalFilter(condition=[IN($7, {
+LogicalProject(DEPTNO=[$7])
+  LogicalFilter(condition=[=($0, +($cor0.EMPNO, 1))])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+})], variablesSet=[[$cor0]])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planMid">
+            <![CDATA[
+LogicalProject(DEPTNO=[$7])
+  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
+    LogicalFilter(condition=[=($7, $9)])
+      LogicalCorrelate(correlation=[$cor0], joinType=[inner], requiredColumns=[{0}])
+        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+        LogicalAggregate(group=[{0}])
+          LogicalProject(DEPTNO=[$7])
+            LogicalFilter(condition=[=($0, +($cor0.EMPNO, 1))])
+              LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+LogicalProject(DEPTNO=[$7])
+  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
+    LogicalJoin(condition=[=($0, +($9, 1))], joinType=[inner])
+      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+      LogicalProject(EMPNO=[$0])
+        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testRemoveSemiJoin2">
+        <Resource name="sql">
+            <![CDATA[select deptno
+from EMP e
+where deptno in (select deptno
+from EMP where empno>e.empno]]>
+        </Resource>
+        <Resource name="planBefore">
+            <![CDATA[
+LogicalProject(DEPTNO=[$7])
+  LogicalFilter(condition=[IN($7, {
+LogicalProject(DEPTNO=[$7])
+  LogicalFilter(condition=[>($0, $cor0.EMPNO)])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+})], variablesSet=[[$cor0]])
+    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planMid">
+            <![CDATA[
+LogicalProject(DEPTNO=[$7])
+  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
+    LogicalFilter(condition=[=($7, $9)])
+      LogicalCorrelate(correlation=[$cor0], joinType=[inner], requiredColumns=[{0}])
+        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+        LogicalAggregate(group=[{0}])
+          LogicalProject(DEPTNO=[$7])
+            LogicalFilter(condition=[>($0, $cor0.EMPNO)])
+              LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+        </Resource>
+        <Resource name="planAfter">
+            <![CDATA[
+LogicalProject(DEPTNO=[$7])
+  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
+    LogicalJoin(condition=[>($0, $9)], joinType=[inner])
+      LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+      LogicalProject(EMPNO=[$0])
+        LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```

Comment:
I am not sure if I missed something, but I did't find the deptno = deptno in the join condition of final plan.

Reply:
Yes you are right, i pruned this condition though deciding the join info left and right keys all come from the same table/view, which is actually like a dim join, Cause SemiJoin eliminates duplicate rows of the output.

Another reason is that this join condition can not be merge into the final join, cause the final join is not a equi join, if we make conjunctions of these join conditions, wrong results will output.

Reply:
It will cause wrong result, won't it?

Reply:
I wound be correct if the join comes from decorrelate, but for common joins(like non semijoin), we should check that the pruned join keys are distinct to the data source, but i make it false as default.

Reply:
Suppose table EMP has 2 rows:
empno  deptno
  1              1
  2              2

Query
```sql
select deptno from EMP e where deptno in (select deptno from EMP where empno>e.empno)
```
should return 0 rows.

How many rows will this plan return?

Reply:
You are right, we should not simplify the non-equi join condition pattern like `a > b`, thus, i think we should make modifications in the RelDecorrelator value generator to recognize the pattern like a = $b + 1, we may change the pattern to $b = a -1 and use a filter instead of the inner join, will do the modifications:

1. remove the SemiJoinRemoveRule
2. Change the RelDecorrelator for join condition pattern `a = $b + 1` `a = $b - 1` `a = $b * 1` `a = $b / 1`